### PR TITLE
[RayService] don't update serveConfigV2 in current ray cluster if ray…

### PR DESCRIPTION
**Pull request by fscnick**
## Why are these changes needed?

When the serveConfigV2 and RayCluster has changed at the same time, it would also update the current active ray cluster with the new serveConfigV2. It might be possible to cause a failure if something in the runtime environment is different.

## Related issue number

Closes https://github.com/ray-project/kuberay/issues/3423

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [ ] Manual tests
  - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by fscnick** [comment]
This is the screenshot of reproduction.

Before applying the fix:
without fix

After applying the fix: (the script periodically check if the new serveConfigV2 has been updated or not until the ray cluster has been terminated.)
with fix

Attach the reproduction script for your reference. 
[issue-3423-reproduce.sh.txt](https://github.com/user-attachments/files/20076295/issue-3423-reproduce.sh.txt)

**Comment by fscnick** [comment]
@kevin85421 PTAL

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 1]
```diff
@@ -82,3 +86,97 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+				strings.Contains(err.Error(), "exit code 6")) {
+			// which is "curl: (52) Empty reply from server" or "curl: (6) Couldn't resolve host"
+			// it implies the ray cluster is terminated.
+			break
+		}
+
+		// Except the errors above, it should not occur.
+		g.Expect(err).NotTo(HaveOccurred())
+
+		g.Expect(stdout.String()).NotTo(ContainSubstring("\"price\": 456"), "new price should not be updated on the old ray cluster")
```
Comment (fscnick) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:180]: There are a few of conditions in this loop.

- check the request to dashboard is unreachable, which implies the ray cluster terminated.  it will exit the loop. (code 52 or 6 might happen)
- check the serveConfigV2 is not contain updated price until the ray cluster terminated.

**Comment by fscnick** [comment]
Hi @kevin85421 , the test is ready. PTAL

**Comment by kevin85421** [comment]
cc @rueian would you mind reviewing this PR? Thanks.

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go (hunk 1 above)
Comment (rueian) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:180]: The implementation looks good to me, but I am afraid that this `g.Expect(stdout.String()).NotTo(ContainSubstring("\"price\": 456"), "new price should not be updated on the old ray cluster")` could never hit due to timing races. Could we simply wait for the old cluster to be destroyed and verify we don't `reconcileServe` on it by checking there are no new `UpdatedServeApplications` or `FailedToUpdateServeApplications` events on the cluster?

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 2]
```diff
@@ -82,3 +86,97 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
 		g.Expect(stdout.String()).To(Equal("9 pizzas please!"))
 	}, TestTimeoutShort).Should(Succeed())
 }
+
+func TestRayServiceInPlaceUpdateWithRayClusterSpec(t *testing.T) {
```
Comment (rueian) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:90]: Can we have another test for zero time upgrade enabled?

Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:90]: It looks like enabled already.

There is no `upgradeStrategy` in [RayServiceSampleYamlApplyConfiguration()](https://github.com/ray-project/kuberay/blob/bf8a9316f094b730e4fcac2090116df701f59e80/ray-operator/test/e2erayservice/support.go#L93)

screenshot:
zero downtime 01
zero downtime 02
zero downtime 03

Comment (rueian) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:89]: Okay, then can we have a test for zero downtime upgrade being disabled?

Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:89]: I'm a bit of confusing about what it should do if zero downtime upgrade is disabled.

If it is disabled, it seems the new RayCluster wouldn't be created. The serveConfigV2 will be update on the existed cluster. Which might be an potential issue that this pr is trying to solve. Because [ShouldPrepareNewCluster(...)](https://github.com/ray-project/kuberay/blob/bf8a9316f094b730e4fcac2090116df701f59e80/ray-operator/controllers/ray/rayservice_controller.go#L273-L279) wouldn't be true under these conditions.

However, if serveConfigV2 is not allow to update on the existed cluster under zero downtime upgrade is disabled with RayClusterSpec and serveConfigV2 modified at the same time. It might violate the design https://docs.ray.io/en/master/cluster/kubernetes/user-guides/rayservice.html#step-7-in-place-update-for-ray-serve-applications.

Comment (rueian) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:89]: > If it is disabled, it seems the new RayCluster wouldn't be created. The serveConfigV2 will be update on the existed cluster.

I think that is the correct behavior. Is there any existing test guarantee that already?

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 3]
```diff
@@ -82,3 +85,95 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+	updateTimestamp := rayService.Status.LastUpdateTime
+
+	// Wait active RayCluster has been terminated
+	g.Eventually(func() string {
+		_, err := GetRayCluster(test, activeRayClusterBeforeUpdate.Namespace, activeRayClusterBeforeUpdate.Name)
+		if err == nil {
+			return ""
+		}
+		return err.Error()
+	}, TestTimeoutMedium).Should(ContainSubstring(fmt.Sprintf("rayclusters.ray.io \"%s\" not found", activeRayClusterBeforeUpdate.Name)))
```
Comment (rueian) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:162]:
```suggestion
	g.Eventually(func() bool {
		_, err := GetRayCluster(test, activeRayClusterBeforeUpdate.Namespace, activeRayClusterBeforeUpdate.Name)
		return k8sApiErrors.IsNotFound(err)
	}, TestTimeoutMedium).Should(BeTrue())
```

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 4]
```diff
@@ -82,3 +85,95 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+
+	events, err := RayServiceEvents(test, rayService)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// event UpdatedServeApplications or FailedToUpdateServeApplications should not occur on activeRayClusterBeforeUpdate after RayService update.
+	for _, event := range events {
+		if event.CreationTimestamp.Before(updateTimestamp) {
```
Comment (rueian) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:169]: shouldn't we test the `LastTimestamp`?

File: ray-operator/test/e2erayservice/support.go [hunk 5]
```diff
@@ -90,6 +90,22 @@ func CurlRayServicePod(
 	return ExecPodCmd(t, curlPod, curlPodContainerName, cmd)
 }
 
+func CurlRayClusterDashboard(
```
Comment (rueian) [inline ray-operator/test/e2erayservice/support.go:93]: I guess this and the ExecPodCmdReturnError can be removed.

File: ray-operator/test/support/events.go [hunk 6]
```diff
@@ -128,3 +131,15 @@ func getWhitespaceStr(size int) string {
 	}
 	return whiteSpaceStr
 }
+
+func RayServiceEvents(t Test, rayService *rayv1.RayService) ([]corev1.Event, error) {
```
Comment (rueian) [inline ray-operator/test/support/events.go:135]: This by itself seems too small to be a helper.

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go (hunk 4 above)
Comment (rueian) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:170]: Yes, LastTimestamp could be later than CreationTimestamp because k8s will aggregate the same event into the old one, therefore if you skip events by the CreationTimestamp, you might skip all the events.

How about you first check if the reason is what we want to test, and then test the LastTimestamp?

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 7]
```diff
@@ -82,3 +87,175 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+
+	rayService, err = test.Client().Ray().RayV1().RayServices(namespace.Name).Update(
+		test.Ctx(),
+		rayService,
+		metav1.UpdateOptions{},
+	)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	updateTimestamp := rayService.Status.LastUpdateTime
+
```
Comment (rueian) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:156]: Add a curl test to make sure the new serve config is updated on the new cluster.

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go (hunk 4 above)
Comment (rueian) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:170]: Oh, the LastTimestamp should record the event we update to the new cluster. How about test the Count = 2?

Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:170]: I'm not sure if it is okay to test the Count == 2. It seems message are excluded from aggregation key so that it is not knowing that belong to which RayCluster. However, if it happens more than 2 or less than 2, it definitely goes wrong.

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go (hunk 7 above)
Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go]: Add it after checking the old RayCluster has terminated.

**Review by rueian — APPROVED** [review]

**Comment by rueian** [comment]
LGTM. cc @kevin85421 for review.

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 8]
```diff
@@ -82,3 +86,177 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+	// event UpdatedServeApplications should occure exactly twice during the entire test.
+	count := 0
+	for _, event := range events.Items {
+		if event.Reason != string(utils.UpdatedServeApplications) {
+			continue
+		}
+		count++
+	}
+	g.Expect(count).To(Equal(2))
```
Comment (kevin85421) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:180]: Can you also check these two events? One is for the old RayCluster, and the other is for the new RayCluster.

**Review by kevin85421 — COMMENTED** [review]
Looks good!

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 9]
```diff
@@ -82,3 +86,177 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+	g.Eventually(RayService(test, rayService.Namespace, rayService.Name), TestTimeoutMedium).
+		Should(WithTransform(IsRayServiceReady, BeTrue()))
+
+	// Get the latest RayService
+	rayService, err = GetRayService(test, namespace.Name, rayServiceName)
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(rayService).NotTo(BeNil())
+
+	// Create curl pod
+	curlPodName := "curl-pod"
```
[the remark this answers is not in the document]
Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:114]: Created. https://github.com/ray-project/kuberay/issues/3613

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go (hunk 8 above)
Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:180]: The RayCluster name appears on `message` which might be exclude from aggregation. It would be difficult to distinguish which is which. There are some discussion about this condition https://github.com/ray-project/kuberay/pull/3559#discussion_r2084908259. Or do I misunderstand something?

Comment (kevin85421) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:175]: In this case, why can we check the number of events?

Comment (kevin85421) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:175]: Discuss offline:

I am OK to merge the PR if:

0. Survey when will aggregation be triggered.
1. It can pass 20 times consecutively in your local env.
2. Check these two events are for different RayCluster CRs
3. Add a comment / TODO for the possibility of aggregation

Comment (kevin85421) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:175]: https://gist.github.com/rueian/fca6c2597ff2ba1dfc2ab5eee148bee7

Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:187]: 0. `Aggregation is performed if a similar event is recorded 10 times in a 10 minute rolling interval.`
https://github.com/kubernetes/client-go/blob/46965213e4561ad1b9c585d1c3551a0cc8d3fcd6/tools/record/events_cache.go#L430-L432

1. Here is the screenshot
rayservice e2e

2. Compare two RayClusters name from the event message at the end.

Comment (kevin85421) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:187]: > Aggregation is performed if a similar event is recorded 10 times in a 10 minute rolling interval.
https://github.com/kubernetes/client-go/blob/46965213e4561ad1b9c585d1c3551a0cc8d3fcd6/tools/record/events_cache.go#L430-L432

Oh, sorry. In that case, aggregation is quite likely to happen since we share the Kubernetes cluster across multiple end-to-end tests. We might need an alternative way to verify that the Ray Serve applications in the old RayCluster are not being updated.

A possible way is:

1. Add a finalizer to old RayCluster
2. Trigger zero downtime upgrade
3. Wait until RayService switches to the new RayCluster
4. Verify the old RayCluster's Serve applications are not updated.
5. Remove the finalizer

Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go]: Fixed. PTAL

**Review by kevin85421 — COMMENTED** [review]
@rueian would you mind reviewing another iteration? I will do an iteration later.

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 10]
```diff
@@ -82,3 +85,198 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+
+	// remove the non-existed-finalizer
+	finalizers := activeRayClusterBeforeUpdate.Finalizers
+	activeRayClusterBeforeUpdate.Finalizers = activeRayClusterBeforeUpdate.Finalizers[:0]
+	for _, finalizer := range finalizers {
+		if finalizer == "no-existed-finalizer" {
+			continue
+		}
+		activeRayClusterBeforeUpdate.Finalizers = append(activeRayClusterBeforeUpdate.Finalizers, finalizer)
+	}
```
Comment (rueian) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:188]:
```suggestion
	activeRayClusterBeforeUpdate.Finalizers = nil
```

Comment (rueian) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:188]: Can't we just clear it?

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 11]
```diff
@@ -82,3 +85,191 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+	// remove the non-existed-finalizer
+	activeRayClusterBeforeUpdate.Finalizers = nil
+	_, err = test.Client().Ray().RayV1().RayClusters(namespace.Name).Update(
+		test.Ctx(),
+		activeRayClusterBeforeUpdate,
+		metav1.UpdateOptions{},
+	)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// make sure the old ray cluster is removed
```
[the remark this answers is not in the document]
Comment (fscnick) [reply ray-operator/test/e2erayservice/rayservice_in_place_update_test.go]: it take elapsed time:  `1m0.446727917s` out of `114.902s`. It might be a bit of long. Remove in advance. Kindly let me know if you'd like to keep it.

File: ray-operator/test/support/core.go [hunk 12]
```diff
@@ -103,6 +103,43 @@ func ExecPodCmd(t Test, pod *corev1.Pod, containerName string, cmd []string) (by
 	return stdout, stderr
 }
 
+func ExecPodCmdReturnError(t Test, pod *corev1.Pod, containerName string, cmd []string) (bytes.Buffer, bytes.Buffer, error) {
```
Comment (kevin85421) [inline ray-operator/test/support/core.go:106]: unused?

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 13]
```diff
@@ -82,3 +83,187 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
 		g.Expect(stdout.String()).To(Equal("9 pizzas please!"))
 	}, TestTimeoutShort).Should(Succeed())
 }
+
+func TestRayServiceInPlaceUpdateWithRayClusterSpec(t *testing.T) {
```
Comment (kevin85421) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:87]: in-place update will not happen in this case.

```suggestion
func TestUpdateServeConfigAndRayClusterSpec(t *testing.T) {
```

File: ray-operator/test/e2erayservice/rayservice_in_place_update_test.go [hunk 14]
```diff
@@ -82,3 +83,187 @@ func TestRayServiceInPlaceUpdate(t *testing.T) {
...
+	oldRayCluster.Finalizers = nil
+	_, err = test.Client().Ray().RayV1().RayClusters(namespace.Name).Update(
+		test.Ctx(),
+		oldRayCluster,
+		metav1.UpdateOptions{},
+	)
+	g.Expect(err).NotTo(HaveOccurred())
+}
+
+func TestRayServiceInPlaceUpdateWithRayClusterSpecWithoutZeroDowntime(t *testing.T) {
```
Comment (kevin85421) [inline ray-operator/test/e2erayservice/rayservice_in_place_update_test.go:189]:
```suggestion
func TestUpdateServeConfigAndRayClusterSpecWithUpgradeDisabled(t *testing.T) {
```

**Review by kevin85421 — COMMENTED** [review]
Overall LGTM

**Review by kevin85421 — APPROVED** [review]
